src/scripts/main.js

```diff
@@ -1,3 +1,446 @@
+      <td class="field__cell" data-position="${rowIndex}-${columnIndex}">
+        ${this.fieldState[rowIndex][columnIndex] || ''}
+      </td>
+    `;
+
+    return tile;
+  }
+  // -- //
+
+  // New Tile Generation
+  generateNewTileValue() {
+    return Math.random() >= 0.9 ? 4 : 2;
+  }
+
+  generateRandomTileIndex() {
+    return Math.floor(Math.random() * this.fieldSize);
+  }
+
+  generateRandomTilePosition() {
+    return [this.generateRandomTileIndex(), this.generateRandomTileIndex()];
+  }
+
+  generateTileInRandomPlace() {
+    if (!this.stateHasEmptyTile()) {
+      return;
+    }
+
+    let emptyTileFound = false;
+
+    while (!emptyTileFound) {
+      const [rowIndex, columnIndex] = this.generateRandomTilePosition();
+
+      if (this.fieldState[rowIndex][columnIndex] === 0) {
+        const tile = this.getTileElementByPosition(rowIndex, columnIndex);
+        const tileValue = this.generateNewTileValue();
+        const { keyframes, options } = this.tileAppearanceAnimation;
+
+        this.fieldState[rowIndex][columnIndex] = tileValue;
+        tile.textContent = tileValue;
+        tile.classList.add(`field__cell--${tileValue}`);
+        tile.animate(keyframes, options);
+        emptyTileFound = true;
+      }
+    }
+  }
+  // -- //
+
+  // Utils
+  isGameWon() {
+    return this.fieldState.flat().includes(2048);
+  }
+
+  isGameLosed() {
+    if (this.stateHasEmptyTile()) {
+      return false;
+    }
+
+    for (let i = 0; i < this.fieldSize; i++) {
+      for (let j = 0; j < this.fieldSize; j++) {
+        const cell = this.fieldState[i][j];
+
+        const cellUnderCurrCell = this.fieldState[i + 1]
+          ? this.fieldState[i + 1][j]
+          : null;
+
+        const rightCellOfCurrCell = this.fieldState[i][j + 1] || null;
+
+        if (cell !== 0) {
+          if (cell === cellUnderCurrCell || cell === rightCellOfCurrCell) {
+            return false;
+          }
+        }
+      }
+    }
+
+    return true;
+  }
+
+  getTileElementByPosition(rowIndex, columnIndex) {
+    return document.querySelector(
+      `[data-position = '${rowIndex + '-' + columnIndex}']`
+    );
+  }
+
+  updateTileByPosition(rowIndex, columnIndex) {
+    const tile = this.getTileElementByPosition(rowIndex, columnIndex);
+    const tileValue = this.fieldState[rowIndex][columnIndex];
+
+    tile.textContent = '';
+    tile.classList.value = 'field__cell';
+
+    if (tileValue > 0) {
+      tile.textContent = tileValue;
+      tile.classList.add('field__cell--' + tileValue);
+    }
+  }
+
+  setScore(value = 0) {
+    const { keyframes, options } = this.scoreRefreshAnimation;
+
+    this.score = value;
+    this.scoreElement.textContent = this.score;
+    this.scoreElement.animate(keyframes, options);
+  }
+
+  getColumnLine(columnIndex) {
+    return [
+      this.fieldState[0][columnIndex],
+      this.fieldState[1][columnIndex],
+      this.fieldState[2][columnIndex],
+      this.fieldState[3][columnIndex],
+    ];
+  }
```

Comment:
Probably in order to make this game field size changable (if we'll want to implement this feature in future) would be better to make something like loop from 0 to `fieldSize - 1` instead of hardcoding, but now it's ok